src/Widgets/PowerModeList.vala

```diff
@@ -0,0 +1,174 @@
+            critical (e.message);
+            successfully_initialized = false;
+        }
+    }
+
+    construct {
+        if (pprofile == null) {
+            return;
+        }
+
+        build_ui ();
+        build_events ();
+    }
+
+    private void build_ui () {
+        Gtk.IconTheme.get_default ().add_resource_path ("/io/elementary/desktop/wingpanel/power");
+
+        var saver_icon = new Gtk.Image.from_icon_name ("power-mode-powersaver-symbolic", Gtk.IconSize.BUTTON);
+
+        var saver_label = new Gtk.Label (_("Power Saver"));
+
+        var saver_button_box = new Gtk.Box (HORIZONTAL, 6) {
+            halign = Gtk.Align.START,
+            margin = 3
+        };
+        saver_button_box.add (saver_icon);
+        saver_button_box.add (saver_label);
+
+        saver_radio = new Gtk.RadioButton (null);
+        saver_radio.get_style_context ().add_class ("image-button");
+        saver_radio.add (saver_button_box);
+
+        var balanced_icon = new Gtk.Image.from_icon_name ("power-mode-balanced-symbolic", Gtk.IconSize.BUTTON);
+
+        var balanced_label = new Gtk.Label (_("Balanced"));
+
+        var balanced_button_box = new Gtk.Box (HORIZONTAL, 6) {
+            halign = Gtk.Align.START,
+            margin = 3
+        };
+        balanced_button_box.add (balanced_icon);
+        balanced_button_box.add (balanced_label);
+
+        balanced_radio = new Gtk.RadioButton.from_widget (saver_radio);
+        balanced_radio.get_style_context ().add_class ("image-button");
+        balanced_radio.add (balanced_button_box);
+
+        var performance_icon = new Gtk.Image.from_icon_name ("power-mode-performance-symbolic", Gtk.IconSize.BUTTON);
+
+        var performance_label = new Gtk.Label (_("Performance"));
+
+        var performance_button_box = new Gtk.Box (HORIZONTAL, 6) {
+            halign = Gtk.Align.START,
+            margin = 3
+        };
+        performance_button_box.add (performance_icon);
+        performance_button_box.add (performance_label);
+
+        performance_radio = new Gtk.RadioButton.from_widget (saver_radio);
+        performance_radio.get_style_context ().add_class ("image-button");
+        performance_radio.add (performance_button_box);
+
+        foreach (unowned var profile in pprofile.profiles) {
+            switch (profile.get ("Profile").get_string ()) {
+                case "power-saver":
+                    add (saver_radio);
+                    break;
+                case "balanced":
+                    add (balanced_radio);
+                    break;
+                case "performance":
+                    add (performance_radio);
+                    break;
+            }
+        }
+    }
+
+    private void build_events () {
+        update_active_profile ();
+
+        saver_radio.toggled.connect (() => {
+            if (saver_radio.active) {
+                settings?.set_string (settings_key, "power-saver");
+                pprofile.active_profile = "power-saver";
```

Comment:
It looks like setting from the settings key does update the backend already, so this should probably only touch the backend if settings is null right?

Reply:
Fixed